### ISSUE TITLE
Add 'test mode' ability

### ISSUE THREAD

**Issue by raar1** [fair-workflows/nanopub#14]
Use known nanopub test servers (at present, only [http://test-server.nanopubs.lod.labs.vu.nl/](http://test-server.nanopubs.lod.labs.vu.nl/)) in order to avoid polluting the existing servers with test publications etc.

**Comment by svenvanderburg** [comment]
@raar1 I am having some trouble figuring out how to use [nanopub-java](https://github.com/Nanopublication/nanopub-java) to publish to the test server. It seems to just publish it to all the servers in a way and there is no documentation on how to test with it (I guess since they only just recently added the test server?) Am I missing something?

**Comment by svenvanderburg** [comment]
Also, it seems like there is no grlc endpoint for the test server: http://grlc.test-server.nanopubs.lod.labs.vu.nl//api/local/local/find_nanopubs_with_text?text=test (this should work, see: http://grlc.nanopubs.lod.labs.vu.nl//api/local/local/find_nanopubs_with_text?text=test)

**Comment by raar1** [comment]
Yes, that's true, there is no grlc endpoint set up for it yet so the search functionality will not yet work with the test servers. I'm hoping we can set it up at some point soon though.

**Comment by tkuhn** [comment]
I could set up the complete nanopub-services package for the test server, if that helps. These services would only return the results from the test server.

**Comment by raar1** [comment]
@tkuhn Yes please, that would be great :)

Returning only the results from the test server would be exactly the behaviour we want.

**Comment by svenvanderburg** [comment]
@tkuhn @raar1 Great! That would allow testing with test grlc endpoints.

We still need a way to publish to the test server though. How to do that?

**Comment by tkuhn** [comment]
OK, here are all the test services:

- http://test-server.nanopubs.lod.labs.vu.nl/
- http://test-grlc.nanopubs.lod.labs.vu.nl/
- http://test-ldf.nanopubs.lod.labs.vu.nl/

To publish a trusty nanopub, just make a POST request with the nanopub as content to to http://test-server.nanopubs.lod.labs.vu.nl/

You might be able to use the `-u` option from the `np publish` command, but now I am not 100% sure anymore whether that will fallback to other servers if the given server fails...

**Comment by svenvanderburg** [comment]
Thanks @tkuhn! I will likely test this this week.

**Comment by svenvanderburg** [comment]
@tkuhn the test grlc endpoints work for us, thanks!

I fail however to publish anything to http://test-server.nanopubs.lod.labs.vu.nl/ using `np publish -u http://test-server.nanopubs.lod.labs.vu.nl/ trigfilename`.
I get:
`Error in nanopub java application: java.lang.RuntimeException: Failed to publish the nanopub`

Using `-v` does not give me much more information:
```
Reading file: /var/folders/81/y875xrcn5_jf7gqry_8s1jv40000gn/T/tmphrrht6ho/signed.temp.trig
---
Trying to publish nanopub: RABMQCtJm5ohyGcCzH_1Jul4Ah4oWO3WB_pB_fYp5fHK8
```

If I use a random different url (e.g. http://random.nl) instead of http://test-server.nanopubs.lod.labs.vu.nl/ it gives me exactly the same error message. If I use http://server.nanopubs.lod.labs.vu.nl/ it does work (sorry for publishing more bogus to the prod servers). This makes me think that the url I am using is wrong.

I am trying to do a post myself (instead of depending on nanopub-java) to see if I get more useful information out of that, but I have a hard time figuring out what that POST should look like. I was wondering if there is any documentation on this? I found the [java code to do it](https://github.com/Nanopublication/nanopub-java/blob/master/src/main/java/org/nanopub/extra/server/PublishNanopub.java#L184), but I don't speak Java that well to reconstruct what I should post exactly.

I hope you can help?

**Comment by tkuhn** [comment]
Ah I see. The nanopub library refuses to publish to a server if the declared public URL doesn't match. And that's how I set up the test server. I have to think of a good solution for this... This works:

    curl -d @test.trig http://test-server.nanopubs.lod.labs.vu.nl/

But I suppose you don't want to introduce a curl dependency.

**Comment by svenvanderburg** [comment]
@tkuhn , ok that explains! Thanks for the curl example, I indeed don't want to introduce a curl dependency, but I should be easily able to do the some thing in python!

**Comment by tkuhn** [comment]
Right, that should indeed be quite straightforward. It would be good if the nanopub library could do that too though. Maybe servers could be explicitly labeled as test servers, but that will require also a bit of change in the server code...

**Comment by svenvanderburg** [comment]
@tkuhn I got it to work! We are populating the [test server](http://test-server.nanopubs.lod.labs.vu.nl/nanopubs.html) with random noise, hooray! Thanks a lot for your active trouble shooting 🙏

**Comment by svenvanderburg** [comment]
Solved by #25, also implemented in fair workflows: https://github.com/fair-workflows/fairworkflows/pull/97

**Comment by tkuhn** [comment]
Awesome! :)

**Comment by svenvanderburg** [comment]
Closing as it is fixed by #25
